### PR TITLE
PS-3706 support native types

### PR DESCRIPTION
**Pull request by pivnicek**
![still draft]
https://keboola.atlassian.net/browse/PS-3706

### CODE REVIEW

File: libs/output-mapping/src/Writer/Table/TableDefinitionColumn.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            if ($type) {
+                $this->dataTypeDefinition = new $nativeTypeClass($type, $options);
+            }
+        }
+        foreach ($metadata as $metadatum) {
+            if ($metadatum['key'] === Common::KBC_METADATA_KEY_BASETYPE) {
+                $this->baseType = $metadatum['value'];
+            }
+        }
+    }
```
Comment (pivnicek) [inline libs/output-mapping/src/Writer/Table/TableDefinitionColumn.php:53]: @odinuv @joker806 I don't like what I've put here so wanted to ping you guys to see if you had maybe some alternative idea of how we can do this :)

File: libs/output-mapping/composer.json [hunk 2]
```diff
@@ -55,7 +56,7 @@
     "scripts": {
         "phplint": "parallel-lint -j 10 --exclude vendor .",
         "phpcs": "phpcs --standard=psr2 --ignore=vendor -n /code",
-        "tests": "phpunit",
+        "tests": "phpunit --filter testCreateTableDefinition",
```
Comment (odinuv) [inline libs/output-mapping/composer.json:59]: Nezapomenout!

File: libs/output-mapping/src/Writer/Table/TableDefinition.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+class TableDefinition
+{
+    public const NATIVE_BACKEND_TYPE_CLASS_MAP = [
+        'snowflake' => Snowflake::class,
+        'synapse' => Synapse::class,
+        'exasol' => Exasol::class,
+    ];
+
+    public const BACKEND_COMPONENTS_MAP = [
```
Comment (odinuv) [inline libs/output-mapping/src/Writer/Table/TableDefinition.php:19]: Toto se mě ani trošku nelibí, z toho se okotíme už při třetím updatu

Comment (pivnicek) [reply libs/output-mapping/src/Writer/Table/TableDefinition.php:19]: jasne, me taky ne

File: libs/output-mapping/src/Writer/TableWriter.php [hunk 4]
```diff
@@ -194,7 +203,21 @@ private function createLoadTableTask(
...
-        if (!$destinationTableExists && $hasColumns) {
+        if ($typedTableEnabled && !$destinationTableExists) {
+            $tableDefinition = (new TableDefinition(
+                $systemMetadata['componentId'],
+                $destinationBucketDetails['backend'])
+            )->setName($destination->getTableName());
+            $tableDefinition->setPrimaryKeysNames(
+                PrimaryKeyHelper::normalizeKeyArray($this->logger, $config['primary_key'])
+            );
+            foreach ($config['column_metadata'] as $columnName => $metadata) {
```
Comment (odinuv) [inline libs/output-mapping/src/Writer/TableWriter.php:214]: this should go into the create table definition method

File: libs/output-mapping/src/Writer/TableWriter.php [hunk 5]
```diff
@@ -239,21 +262,30 @@ private function createLoadTableTask(
...
         $destinationBucketId = $destination->getBucketId();
-        $destinationBucketExists = $this->clientWrapper->getBasicClient()->bucketExists($destinationBucketId);
+        try {
+            $destinationBucketDetails = $this->clientWrapper->getBasicClient()->getBucket($destinationBucketId);
+            $destinationBucketExists = true;
+        } catch (ClientException $e) {
+            if ($e->getCode() == 404) {
+                $destinationBucketExists = false;
+            }
+            throw $e;
```
Comment (odinuv) [inline libs/output-mapping/src/Writer/TableWriter.php:275]: why?

File: libs/output-mapping/src/Writer/TableWriter.php (hunk 4 above)
Comment (pivnicek) [reply libs/output-mapping/src/Writer/TableWriter.php:214]: yup, I'm almost finished with my refactoring, using factories

File: libs/output-mapping/src/Writer/TableWriter.php (hunk 5 above)
Comment (pivnicek) [reply libs/output-mapping/src/Writer/TableWriter.php:275]: why is it thrown?  it was being thrown before in the `bucketExists` method

File: libs/output-mapping/src/Writer/TableWriter.php [hunk 6]
```diff
@@ -239,21 +262,30 @@ private function createLoadTableTask(
         return $loadTask;
     }
 
-    private function ensureValidDestinationBucketExists(MappingDestination $destination, array $systemMetadata)
+    private function ensureValidDestinationBucketExists(MappingDestination $destination, array $systemMetadata): array
```
Comment (odinuv) [inline libs/output-mapping/src/Writer/TableWriter.php:265]: should be renamed

File: libs/output-mapping/src/Writer/Table/TableDefinition.php (hunk 3 above)
Comment (odinuv) [reply libs/output-mapping/src/Writer/Table/TableDefinition.php:19]: let's avoid it then

File: libs/output-mapping/src/Writer/TableWriter.php (hunk 5 above)
Comment (odinuv) [reply libs/output-mapping/src/Writer/TableWriter.php:275]: no it wasn't - that doesn't make sense, why would it set the $destinationBucketExists variable then?

Comment (pivnicek) [reply libs/output-mapping/src/Writer/TableWriter.php:275]: ah fuck, right!  it returns there 🤦 .  thanks odin

File: libs/output-mapping/src/Writer/Table/TableDefinition.php (hunk 3 above)
Comment (pivnicek) [reply libs/output-mapping/src/Writer/Table/TableDefinition.php]: yeah, I'm just not quite sure how 😄 .  it's now been moved to the factory class

**Comment by ErikZigo** [comment]
@pivnicek what is status for that? is it ready for another review round? Odin will have vacation tomorrow, so I expect I should take over it.

**Comment by pivnicek** [comment]
@ErikZigo I just finished a call with Odin and will change this tonight.  It'll be ready for your review in the morning, thanks!

btw, it will be greatly simplified :-)

**Comment by ErikZigo** [comment]
I expect you are on Czech Language lesson :)

**Comment by pivnicek** [comment]
jsou prazdniny 😄

**Comment by odinuv** [comment]
greatly simplified means that it will work with base types and not with the native types yet (because that is a subset of cases with the base types)

**Comment by pivnicek** [comment]
closing because it was solved in other PRs

**Pull request closed without merging**